### DISCUSSION TITLE
New architecture support [LoongArch]

### DISCUSSION THREAD

**Discussion by yzewei** [pypa/manylinux#1587] [Ideas]
[BackGround]
LoongArch is a new RISC ISA, which is independently designed by Loongson Technology.

LoongArch includes a reduced 32-bit version (LA32R), a standard 32-bit version (LA32S) and a 64-bit version (LA64), and loong64 is the 64-bit version of LoongArch.

LoongArch documentation: https://github.com/loongson/LoongArch-Documentation.git

- Alpine support:  https://alpinelinux.org/posts/Alpine-3.21.0-released.html

- Debian sid:  https://wiki.debian.org/ArchiveQualification/loong64

- docker buildx already support loongarch64 in x86_64 platform

[What do I want to do]

Provide the ci pipeline build part of loongarch

**Comment by yzewei** [comment]
@alex

**Comment by mayeut** [comment]
The same answers as #1426 apply here.
We'd need support from a base image. Ideally, RHEL (or a derivative like AlmaLinux/RockyLinux). However another base image might be okay for loongarch only.
Native CI availability is a must-have as qemu is a no go here as it is way to slow to build images (we were often hitting GHA time limit on ppc64le when travis-ci was broken on that architecture).

**Reply by yzewei** [reply]
Is it possible to use docker buildx?
Only qemu-user mode is applied

**Reply by mayeut** [reply]
This is what the ppc64le build was using on  GHA. It's too slow.

**Reply by snnn** [reply]
Cross-compiling might be a possible solution. CPython's build system is autoconf/automake based, which supports cross-compiling very well. In theory we can get a LoongArch base image from somewhere, then cross-compiling all python versions for it, then copy the files into the image then publish the image. If just Python needs be handled in this way, it wouldn't be too hard. However, you may also need to build gcc/llvm from scratch, where you probably will also hit the time limit. 
You cannot wait Github or Azure to have LoongArch available because cloud platforms only use very popular commodity available hardware. (This my personal opinion, doesn't represent the company). 
On the other hand, Python is an open community. Nobody says you must use RHEL.  You can build your wheels from any Linux distro then publish them to pypi as long as the tag you chose can be recognized by pip and auditwheel.

**Comment by yzewei** [comment]
"I've been following the discussions and pull requests concerning new architecture support for manylinux, particularly at https://github.com/pypa/manylinux/discussions/1426. Since s390x and RISC-V builds are feasible via emulation, I wanted to inquire about the possibility of adding support for LoongArch. This architecture has recently gained official support in Alpine Linux, and Debian's sid branch already provides packages for it.

It's also worth noting that we can readily perform builds using the loongarch64/debian Docker image available at https://hub.docker.com/r/loongarch64/debian. This process is indeed feasible, and we even maintain manylinux image artifacts in our unofficial Docker image repository: https://lcr.loongnix.cn/repository/pypa/manylinux_2_38_loongarch64. Any consideration for LoongArch would be greatly appreciated!"

**Reply by mayeut** [reply]
Debian sid is a no-go so this needs to wait at least for the next release. While Debian can be used it's not usually a good base for manylinux (but this has been done for armv7l where there are no RHEL derivatives available).
There are no official images for alpine on [Docker Hub](https://hub.docker.com/_/alpine) and the one in https://hub.docker.com/r/loongarch64/alpine is outdated which would result in inconsistencies in the musllinux_1_2 image and/or added maintenance burden.

The images here are using native packages published on PyPI so publication on PyPI should be allowed and, as the discussion around riscv64 showed, it can take some time.

**Reply by yzewei** [reply]
> Debian sid is a no-go so this needs to wait at least for the next release. While Debian can be used it's not usually a good base for manylinux (but this has been done for armv7l where there are no RHEL derivatives available). There are no official images for alpine on [Docker Hub](https://hub.docker.com/_/alpine) and the one in https://hub.docker.com/r/loongarch64/alpine is outdated which would result in inconsistencies in the musllinux_1_2 image and/or added maintenance burden.
> 
> The images here are using native packages published on PyPI so publication on PyPI should be allowed and, as the discussion around riscv64 showed, it can take some time.

OK. Wait for next Release.

**Reply by cheungxi** [reply]
> Debian sid is a no-go so this needs to wait at least for the next release. While Debian can be used it's not usually a good base for manylinux (but this has been done for armv7l where there are no RHEL derivatives available). There are no official images for alpine on [Docker Hub](https://hub.docker.com/_/alpine) and the one in https://hub.docker.com/r/loongarch64/alpine is outdated which would result in inconsistencies in the musllinux_1_2 image and/or added maintenance burden.
> 
> The images here are using native packages published on PyPI so publication on PyPI should be allowed and, as the discussion around riscv64 showed, it can take some time.

Now that PR #1927 has been merged and OpenAnolis 23.4 / Alpine 3.22 have been selected as the base images, what are the remaining blockers to officially building and publishing the LoongArch64 images?

Is native LoongArch64 CI availability currently the main blocker, or are there still unresolved concerns about the base images, build process, or maintenance requirements?

Is anyone already working on the CI or image publishing side?

Would it be useful to open a separate issue to track enabling the LoongArch64 build matrix and publishing the official images?

**Comment by mayeut** [comment]
The remains blocker is, as written in https://github.com/pypa/manylinux/pull/1927: publication allowed in PyPI.
The status changed here for native CI since we are now cross-compiling and not hitting any limits.
